File: implementation/scipamato/core/pubmed-api/src/test/java/ch/difty/scipamato/core/pubmed/ScipamatoPubmedArticleTest.java
package ch.difty.scipamato.core.pubmed;

import static ch.difty.scipamato.common.TestUtilsKt.assertDegenerateSupplierParameter;
import static org.assertj.core.api.Assertions.assertThat;

import java.lang.Object;
import java.util.List;

import nl.jqno.equalsverifier.EqualsVerifier;
import nl.jqno.equalsverifier.Warning;
import org.junit.jupiter.api.Test;

import ch.difty.scipamato.core.pubmed.api.*;

class ScipamatoPubmedArticleTest {

    private final PubmedArticle pubmedArticle = makeMinimalValidPubmedArticle();

    static PubmedArticle makeMinimalValidPubmedArticle() {
        PubmedArticle pa = new PubmedArticle();
        MedlineCitation medlineCitation = new MedlineCitation();
        Article article = new Article();
        Journal journal = new Journal();
        JournalIssue journalIssue = new JournalIssue();
        journalIssue.setPubDate(new PubDate());
        journal.setJournalIssue(journalIssue);
        article.setJournal(journal);
        article.setArticleTitle(new ArticleTitle());
        medlineCitation.setArticle(article);
        medlineCitation.setPMID(new PMID());
        medlineCitation.setMedlineJournalInfo(new MedlineJournalInfo());
        pa.setMedlineCitation(medlineCitation);
        return pa;
    }

    @Test
    void degenerateConstruction_withNullPubmedArticle() {
        assertDegenerateSupplierParameter(() -> new ScipamatoPubmedArticle(null), "pubmedArticle");
    }

    @Test
    void degenerateConstruction_withNullMedlineCitation() {
        pubmedArticle.setMedlineCitation(null);
        assertDegenerateSupplierParameter(() -> new ScipamatoPubmedArticle(pubmedArticle),
            "pubmedArticle.medlineCitation");
    }

    @Test
    void degenerateConstruction_withNullArticle() {
        pubmedArticle
            .getMedlineCitation()
            .setArticle(null);
        assertDegenerateSupplierParameter(() -> new ScipamatoPubmedArticle(pubmedArticle),
            "pubmedArticle.medlineCitation.article");
    }

    @Test
    void degenerateConstruction_withNullJournal() {
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .setJournal(null);
        assertDegenerateSupplierParameter(() -> new ScipamatoPubmedArticle(pubmedArticle),
            "pubmedArticle.medlineCitation.article.journal");
    }

    @Test
    void degenerateConstruction_withNullPmid() {
        pubmedArticle
            .getMedlineCitation()
            .setPMID(null);
        assertDegenerateSupplierParameter(() -> new ScipamatoPubmedArticle(pubmedArticle),
            "pubmedArticle.medlineCitation.pmid");
    }

    @Test
    void degenerateConstruction_withNullJournalIssue() {
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .setJournalIssue(null);
        assertDegenerateSupplierParameter(() -> new ScipamatoPubmedArticle(pubmedArticle),
            "pubmedArticle.medlineCitation.article.journal.journalIssue");
    }

    @Test
    void degenerateConstruction_withNullPubDate() {
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .setPubDate(null);
        assertDegenerateSupplierParameter(() -> new ScipamatoPubmedArticle(pubmedArticle),
            "pubmedArticle.medlineCitation.article.journal.journalIssue.pubDate");
    }

    @Test
    void degenerateConstruction_withNullMedlineJournalInfo() {
        pubmedArticle
            .getMedlineCitation()
            .setMedlineJournalInfo(null);
        assertDegenerateSupplierParameter(() -> new ScipamatoPubmedArticle(pubmedArticle),
            "pubmedArticle.medlineCitation.medlineJournalInfo");
    }

    @Test
    void validConstruction() {
        assertThat(new ScipamatoPubmedArticle(pubmedArticle)).isNotNull();
    }

    @Test
    void validConstructionUsingOf() {
        assertThat(PubmedArticleFacade.newPubmedArticleFrom(pubmedArticle)).isNotNull();
    }

    @Test
    void extractingYearFromNeitherYearObjectNorMedlineDate_returnsYear0() {
        assertThat(pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .getPubDate()
            .getYearOrMonthOrDayOrSeasonOrMedlineDate()).isEmpty();
        Month month = new Month();
        month.setvalue("2016");
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .getPubDate()
            .getYearOrMonthOrDayOrSeasonOrMedlineDate()
            .add(month);
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getPublicationYear()).isEqualTo("0");
    }

    @Test
    void extractYearFromYearObject() {
        Year year = new Year();
        year.setvalue("2016");
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .getPubDate()
            .getYearOrMonthOrDayOrSeasonOrMedlineDate()
            .add(year);
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getPublicationYear()).isEqualTo("2016");
    }

    @Test
    void extractYearFromMedlineDate() {
        MedlineDate md = new MedlineDate();
        md.setvalue("2016 Nov-Dec");
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .getPubDate()
            .getYearOrMonthOrDayOrSeasonOrMedlineDate()
            .add(md);

        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getPublicationYear()).isEqualTo("2016");
    }

    @Test
    void withNoFurtherAttributes() {
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getAuthors()).isNull();
        assertThat(spa.getFirstAuthor()).isNull();
        assertThat(spa.getLocation()).isEqualTo("null. 0;");
        assertThat(spa.getTitle()).isNull();
        assertThat(spa.getDoi()).isNull();
    }

    @Test
    void authors_withoutCollectiveAuthor() {
        AuthorList authorList = new AuthorList();
        authorList
            .getAuthor()
            .add(makeAuthor("Bond", "James", "J", null, null));
        authorList
            .getAuthor()
            .add(makeAuthor("Kid", "Billy", "B", "Jr", null));
        authorList
            .getAuthor()
            .add(makeAuthor("Joice", "James", "J", null, null));
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .setAuthorList(authorList);
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getAuthors()).isEqualTo("Bond J, Kid B Jr, Joice J.");
        assertThat(spa.getFirstAuthor()).isEqualTo("Bond");
    }

    @Test
    void authors_withCollectiveAuthor() {
        AuthorList authorList = new AuthorList();
        authorList
            .getAuthor()
            .add(makeAuthor("Bond", "James", "J", null, null));
        authorList
            .getAuthor()
            .add(makeAuthor("Kid", "Billy", "B", "Jr", null));
        authorList
            .getAuthor()
            .add(makeAuthor(null, null, null, null, "Joice J"));
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .setAuthorList(authorList);
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getAuthors()).isEqualTo("Bond J, Kid B Jr; Joice J.");
        assertThat(spa.getFirstAuthor()).isEqualTo("Bond");
    }

    @Test
    void noAuthors_withCollectiveAuthor() {
        AuthorList authorList = new AuthorList();
        authorList
            .getAuthor()
            .add(makeAuthor(null, null, null, null, "Joice J"));
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .setAuthorList(authorList);
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getAuthors()).isEqualTo("Joice J.");
        assertThat(spa.getFirstAuthor()).isEqualTo("Joice J");
    }

    @Test
    void noAuthors_noCollectiveAuthor() {
        AuthorList authorList = new AuthorList();
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .setAuthorList(authorList);
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getAuthors()).isEqualTo("");
        assertThat(spa.getFirstAuthor()).isEqualTo("");
    }

    @Test
    void location_withMedlinePagination() {
        pubmedArticle
            .getMedlineCitation()
            .getMedlineJournalInfo()
            .setMedlineTA("Medline TA");
        Year year = new Year();
        year.setvalue("2016");
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .getPubDate()
            .getYearOrMonthOrDayOrSeasonOrMedlineDate()
            .add(year);
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .setVolume("6");
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .setIssue("10");
        Pagination pagination = new Pagination();
        MedlinePgn medlinePgn = new MedlinePgn();
        medlinePgn.setvalue("1145-9");
        pagination
            .getStartPageOrEndPageOrMedlinePgn()
            .add(medlinePgn);
        MedlinePgn medlinePgn2 = new MedlinePgn();
        medlinePgn2.setvalue("3456-3458");
        pagination
            .getStartPageOrEndPageOrMedlinePgn()
            .add(medlinePgn2);
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getPaginationOrELocationID()
            .add(pagination);
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getLocation()).isEqualTo("Medline TA. 2016; 6 (10): 1145-1149.");
    }

    @Test
    void location_withMedlinePaginationWithoutPageRange() {
        pubmedArticle
            .getMedlineCitation()
            .getMedlineJournalInfo()
            .setMedlineTA("Medline TA");
        Year year = new Year();
        year.setvalue("2016");
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .getPubDate()
            .getYearOrMonthOrDayOrSeasonOrMedlineDate()
            .add(year);
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .setVolume("6");
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .setIssue("10");
        Pagination pagination = new Pagination();
        MedlinePgn medlinePgn = new MedlinePgn();
        medlinePgn.setvalue("1145");
        pagination
            .getStartPageOrEndPageOrMedlinePgn()
            .add(medlinePgn);
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getPaginationOrELocationID()
            .add(pagination);
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getLocation()).isEqualTo("Medline TA. 2016; 6 (10): 1145.");
    }

    @Test
    void location_withMedlineElocation() {
        pubmedArticle
            .getMedlineCitation()
            .getMedlineJournalInfo()
            .setMedlineTA("Medline TA");
        Year year = new Year();
        year.setvalue("2016");
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .getPubDate()
            .getYearOrMonthOrDayOrSeasonOrMedlineDate()
            .add(year);
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .setVolume("6");
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getJournal()
            .getJournalIssue()
            .setIssue("10");
        ELocationID elocationId1 = new ELocationID();
        elocationId1.setEIdType("foo");
        elocationId1.setvalue("bar");
        ELocationID elocationId2 = new ELocationID();
        elocationId2.setEIdType("pii");
        elocationId2.setvalue("baz");
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getPaginationOrELocationID()
            .add(elocationId1);
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getPaginationOrELocationID()
            .add(elocationId2);
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getLocation()).isEqualTo("Medline TA. 2016; 6 (10). pii: baz.");
    }

    private Author makeAuthor(String lastname, String forename, String initials, String suffix, String collectiveName) {
        Author author = new Author();
        if (lastname != null) {
            LastName ln = new LastName();
            ln.setvalue(lastname);
            author
                .getLastNameOrForeNameOrInitialsOrSuffixOrCollectiveName()
                .add(ln);
        }
        if (forename != null) {
            ForeName fn = new ForeName();
            fn.setvalue(forename);
            author
                .getLastNameOrForeNameOrInitialsOrSuffixOrCollectiveName()
                .add(fn);
        }
        if (initials != null) {
            Initials i = new Initials();
            i.setvalue(initials);
            author
                .getLastNameOrForeNameOrInitialsOrSuffixOrCollectiveName()
                .add(i);
        }
        if (suffix != null) {
            Suffix s = new Suffix();
            s.setvalue(suffix);
            author
                .getLastNameOrForeNameOrInitialsOrSuffixOrCollectiveName()
                .add(s);
        }
        if (collectiveName != null) {
            CollectiveName cn = new CollectiveName();
            cn.setvalue(collectiveName);
            author
                .getLastNameOrForeNameOrInitialsOrSuffixOrCollectiveName()
                .add(cn);
        }
        return author;
    }

    @Test
    void title() {
        ArticleTitle articleTitle = new ArticleTitle();
        articleTitle.setMixedContent(List.of("article title"));
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .setArticleTitle(articleTitle);
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getTitle()).isEqualTo("article title");
    }

    @Test
    void doi_fromArticleIdList() {
        ArticleId articleId1 = new ArticleId();
        articleId1.setIdType("foo");
        articleId1.setvalue("bar");
        ArticleId articleId2 = new ArticleId();
        articleId2.setIdType("doi");
        articleId2.setvalue("10.0000012345");
        ArticleIdList articleIdList = new ArticleIdList();
        articleIdList
            .getArticleId()
            .add(articleId1);
        articleIdList
            .getArticleId()
            .add(articleId2);
        PubmedData pubmedData = new PubmedData();
        pubmedData.setArticleIdList(articleIdList);
        pubmedArticle.setPubmedData(pubmedData);
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getDoi()).isEqualTo("10.0000012345");
    }

    @Test
    void doi_withoutArticleIdList_usesElocIfthere() {
        ELocationID elocationId = new ELocationID();
        elocationId.setvalue("eloc");
        pubmedArticle
            .getMedlineCitation()
            .getArticle()
            .getPaginationOrELocationID()
            .add(elocationId);
        ScipamatoPubmedArticle spa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(spa.getDoi()).isEqualTo("eloc");
    }

    @Test
    void equals() {
        EqualsVerifier
            .forClass(ScipamatoPubmedArticle.class)
            .suppress(Warning.STRICT_INHERITANCE, Warning.NONFINAL_FIELDS)
            .verify();
    }

    @Test
    void testingToString() {
        ScipamatoPubmedArticle pa = new ScipamatoPubmedArticle(pubmedArticle);
        assertThat(pa.toString()).isEqualTo(
            "AbstractPubmedArticleFacade(pmId=null, authors=null, firstAuthor=null, publicationYear=0, location=null. 0;, title=null, doi=null, originalAbstract=null)");
    }

    @Test
    void handleDatishObject_withYear() {
        Year o = new Year();
        o.setvalue("2018");
        assertDatish("2018 ", o);
    }

    @Test
    void handleDatishObject_withMonth() {
        Month o = new Month();
        o.setvalue("12");
        assertDatish("12 ", o);
    }

    @Test
    void handleDatishObject_withDay() {
        Day o = new Day();
        o.setvalue("13");
        assertDatish("13", o);
    }

    @Test
    void handleDatishObject_withUndefinedObject() {
        assertDatish("", "foo");
    }

    private void assertDatish(final String expected, final Object o) {
        final ScipamatoPubmedArticle pa = new ScipamatoPubmedArticle(pubmedArticle);
        final StringBuilder sb = new StringBuilder();
        pa.handleDatishObject(sb, o);
        assertThat(sb.toString()).isEqualTo(expected);
    }
}

File: implementation/scipamato/core/pubmed-api/src/test/kotlin/ch/difty/scipamato/core/pubmed/ScipamatoPubmedArticleTest.kt
package ch.difty.scipamato.core.pubmed

import ch.difty.scipamato.core.pubmed.api.*
import nl.jqno.equalsverifier.EqualsVerifier
import nl.jqno.equalsverifier.Warning
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

@Suppress("SpellCheckingInspection")
internal class ScipamatoPubmedArticleTest {

    private val pubmedArticle = makeMinimalValidPubmedArticle()

    @Test
    fun extractingYearFromNeitherYearObjectNorMedlineDate_returnsYear0() {
        assertThat(pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue
            .pubDate
            .yearOrMonthOrDayOrSeasonOrMedlineDate).isEmpty()
        val month = Month()
        month.setvalue("2016")
        pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue
            .pubDate
            .yearOrMonthOrDayOrSeasonOrMedlineDate
            .add(month)
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.publicationYear).isEqualTo("0")
    }

    @Test
    fun extractYearFromYearObject() {
        val year = Year()
        year.setvalue("2016")
        pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue
            .pubDate
            .yearOrMonthOrDayOrSeasonOrMedlineDate
            .add(year)
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.publicationYear).isEqualTo("2016")
    }

    @Test
    fun extractYearFromMedlineDate() {
        val md = MedlineDate()
        md.setvalue("2016 Nov-Dec")
        pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue
            .pubDate
            .yearOrMonthOrDayOrSeasonOrMedlineDate
            .add(md)

        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.publicationYear).isEqualTo("2016")
    }

    @Test
    fun withNoFurtherAttributes() {
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.authors).isNull()
        assertThat(spa.firstAuthor).isNull()
        assertThat(spa.location).isEqualTo("null. 0;")
        assertThat(spa.title).isNull()
        assertThat(spa.doi).isNull()
    }

    @Test
    fun authors_withoutCollectiveAuthor() {
        val authorList = AuthorList()
        authorList.author.add(makeAuthor("Bond", "James", "J", null, null))
        authorList.author.add(makeAuthor("Kid", "Billy", "B", "Jr", null))
        authorList.author.add(makeAuthor("Joice", "James", "J", null, null))
        pubmedArticle.medlineCitation.article.authorList = authorList
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.authors).isEqualTo("Bond J, Kid B Jr, Joice J.")
        assertThat(spa.firstAuthor).isEqualTo("Bond")
    }

    @Test
    fun authors_withCollectiveAuthor() {
        val authorList = AuthorList()
        authorList.author.add(makeAuthor("Bond", "James", "J", null, null))
        authorList.author.add(makeAuthor("Kid", "Billy", "B", "Jr", null))
        authorList.author.add(makeAuthor(null, null, null, null, "Joice J"))
        pubmedArticle.medlineCitation.article.authorList = authorList
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.authors).isEqualTo("Bond J, Kid B Jr; Joice J.")
        assertThat(spa.firstAuthor).isEqualTo("Bond")
    }

    @Test
    fun noAuthors_withCollectiveAuthor() {
        val authorList = AuthorList()
        authorList.author.add(makeAuthor(null, null, null, null, "Joice J"))
        pubmedArticle.medlineCitation.article.authorList = authorList
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.authors).isEqualTo("Joice J.")
        assertThat(spa.firstAuthor).isEqualTo("Joice J")
    }

    @Test
    fun noAuthors_noCollectiveAuthor() {
        val authorList = AuthorList()
        pubmedArticle.medlineCitation.article.authorList = authorList
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.authors).isEqualTo("")
        assertThat(spa.firstAuthor).isEqualTo("")
    }

    @Test
    fun location_withMedlinePagination() {
        pubmedArticle
            .medlineCitation
            .medlineJournalInfo.medlineTA = "Medline TA"
        val year = Year()
        year.setvalue("2016")
        pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue
            .pubDate
            .yearOrMonthOrDayOrSeasonOrMedlineDate
            .add(year)
        pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue.volume = "6"
        pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue.issue = "10"
        val pagination = Pagination()
        val medlinePgn = MedlinePgn()
        medlinePgn.setvalue("1145-9")
        pagination
            .startPageOrEndPageOrMedlinePgn
            .add(medlinePgn)
        val medlinePgn2 = MedlinePgn()
        medlinePgn2.setvalue("3456-3458")
        pagination
            .startPageOrEndPageOrMedlinePgn
            .add(medlinePgn2)
        pubmedArticle
            .medlineCitation
            .article
            .paginationOrELocationID
            .add(pagination)
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.location).isEqualTo("Medline TA. 2016; 6 (10): 1145-1149.")
    }

    @Test
    fun location_withMedlinePaginationWithoutPageRange() {
        pubmedArticle
            .medlineCitation
            .medlineJournalInfo.medlineTA = "Medline TA"
        val year = Year()
        year.setvalue("2016")
        pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue
            .pubDate
            .yearOrMonthOrDayOrSeasonOrMedlineDate
            .add(year)
        pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue.volume = "6"
        pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue.issue = "10"
        val pagination = Pagination()
        val medlinePgn = MedlinePgn()
        medlinePgn.setvalue("1145")
        pagination
            .startPageOrEndPageOrMedlinePgn
            .add(medlinePgn)
        pubmedArticle
            .medlineCitation
            .article
            .paginationOrELocationID
            .add(pagination)
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.location).isEqualTo("Medline TA. 2016; 6 (10): 1145.")
    }

    @Test
    fun location_withMedlineElocation() {
        pubmedArticle.medlineCitation.medlineJournalInfo.medlineTA = "Medline TA"
        val year = Year()
        year.setvalue("2016")
        pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue
            .pubDate
            .yearOrMonthOrDayOrSeasonOrMedlineDate
            .add(year)
        pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue.volume = "6"
        pubmedArticle
            .medlineCitation
            .article
            .journal
            .journalIssue.issue = "10"
        val elocationId1 = ELocationID()
        elocationId1.eIdType = "foo"
        elocationId1.setvalue("bar")
        val elocationId2 = ELocationID()
        elocationId2.eIdType = "pii"
        elocationId2.setvalue("baz")
        pubmedArticle
            .medlineCitation
            .article
            .paginationOrELocationID
            .add(elocationId1)
        pubmedArticle
            .medlineCitation
            .article
            .paginationOrELocationID
            .add(elocationId2)
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.location).isEqualTo("Medline TA. 2016; 6 (10). pii: baz.")
    }

    private fun makeAuthor(
        lastname: String?,
        forename: String?,
        initials: String?,
        suffix: String?,
        collectiveName: String?
    ): Author {
        val author = Author()
        if (lastname != null) {
            val ln = LastName()
            ln.setvalue(lastname)
            author.lastNameOrForeNameOrInitialsOrSuffixOrCollectiveName.add(ln)
        }
        if (forename != null) {
            val fn = ForeName()
            fn.setvalue(forename)
            author.lastNameOrForeNameOrInitialsOrSuffixOrCollectiveName.add(fn)
        }
        if (initials != null) {
            val i = Initials()
            i.setvalue(initials)
            author.lastNameOrForeNameOrInitialsOrSuffixOrCollectiveName.add(i)
        }
        if (suffix != null) {
            val s = Suffix()
            s.setvalue(suffix)
            author.lastNameOrForeNameOrInitialsOrSuffixOrCollectiveName.add(s)
        }
        if (collectiveName != null) {
            val cn = CollectiveName()
            cn.setvalue(collectiveName)
            author.lastNameOrForeNameOrInitialsOrSuffixOrCollectiveName.add(cn)
        }
        return author
    }

    @Test
    fun title() {
        val articleTitle = ArticleTitle()
        articleTitle.mixedContent = listOf("article title")
        pubmedArticle.medlineCitation.article.articleTitle = articleTitle
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.title).isEqualTo("article title")
    }

    @Test
    fun doi_fromArticleIdList() {
        val articleId1 = ArticleId()
        articleId1.idType = "foo"
        articleId1.setvalue("bar")
        val articleId2 = ArticleId()
        articleId2.idType = "doi"
        articleId2.setvalue("10.0000012345")
        val articleIdList = ArticleIdList()
        articleIdList.articleId.add(articleId1)
        articleIdList.articleId.add(articleId2)
        val pubmedData = PubmedData()
        pubmedData.articleIdList = articleIdList
        pubmedArticle.pubmedData = pubmedData
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.doi).isEqualTo("10.0000012345")
    }

    @Test
    fun doi_withoutArticleIdList_usesElocIfthere() {
        val elocationId = ELocationID()
        elocationId.setvalue("eloc")
        pubmedArticle
            .medlineCitation
            .article
            .paginationOrELocationID
            .add(elocationId)
        val spa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(spa.doi).isEqualTo("eloc")
    }

    @Test
    fun equals() {
        EqualsVerifier
            .forClass(ScipamatoPubmedArticle::class.java)
            .suppress(Warning.STRICT_INHERITANCE, Warning.NONFINAL_FIELDS)
            .verify()
    }

    @Test
    fun testingToString() {
        val pa = ScipamatoPubmedArticle(pubmedArticle)
        assertThat(pa.toString()).isEqualTo(
            "AbstractPubmedArticleFacade(pmId=null, authors=null, firstAuthor=null, publicationYear=0, " +
                "location=null. 0;, title=null, doi=null, originalAbstract=null)"
        )
    }

    @Test
    fun handleDatishObject_withYear() {
        val o = Year()
        o.setvalue("2018")
        assertDatish("2018 ", o)
    }

    @Test
    fun handleDatishObject_withMonth() {
        val o = Month()
        o.setvalue("12")
        assertDatish("12 ", o)
    }

    @Test
    fun handleDatishObject_withDay() {
        val o = Day()
        o.setvalue("13")
        assertDatish("13", o)
    }

    @Test
    fun handleDatishObject_withUndefinedObject() {
        assertDatish("", "foo")
    }

    private fun assertDatish(expected: String, o: Any) {
        val pa = ScipamatoPubmedArticle(pubmedArticle)
        val sb = StringBuilder()
        pa.handleDatishObject(sb, o)
        assertThat(sb.toString()).isEqualTo(expected)
    }

    companion object {

        fun makeMinimalValidPubmedArticle(): PubmedArticle {
            val pa = PubmedArticle()
            val medlineCitation = MedlineCitation()
            val article = Article()
            val journal = Journal()
            val journalIssue = JournalIssue()
            journalIssue.pubDate = PubDate()
            journal.journalIssue = journalIssue
            article.journal = journal
            article.articleTitle = ArticleTitle()
            medlineCitation.article = article
            medlineCitation.pmid = PMID()
            medlineCitation.medlineJournalInfo = MedlineJournalInfo()
            pa.medlineCitation = medlineCitation
            return pa
        }
    }
}
